Mathlib/CategoryTheory/Shift/ShiftedHom.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Shift.Basic
+import Mathlib.CategoryTheory.Preadditive.AdditiveFunctor
+
+/-! Shifted morphisms
+
+Given a category `C` endowed with a shift by an additive monoid `M` and two
+objects `X` and `Y` in `C`, we consider the types `ShiftedHom X Y m`
+defined as `X ⟶ (Y⟦m⟧)` for all `m : M`, and the composition on these
+shifted hom.
+
+## TODO
+
+* redefine Ext-groups in abelian categories using `ShiftedHom` in the derived category.
+* study the `R`-module structures on `ShiftedHom` when `C` is `R`-linear
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+variable {C : Type*} [Category C]
+  {M : Type*} [AddMonoid M] [HasShift C M]
+
+/-- In a category `C` equipped with a shift by an additive monoid,
+this is the type of morphisms `X ⟶ (Y⟦n⟧)` for `m : M`.  -/
+def ShiftedHom (X Y : C) (m : M) : Type _ := X ⟶ (Y⟦m⟧)
+
+instance [Preadditive C] (X Y : C) (n : M) : AddCommGroup (ShiftedHom X Y n) := by
+  dsimp only [ShiftedHom]
+  infer_instance
+
+namespace ShiftedHom
+
+variable {X Y Z T : C}
+
+/-- The composition of `f : X ⟶ Y⟦a⟧` and `g : Y ⟶ Z⟦b⟧`, as a morphism `X ⟶ Z⟦c⟧`
+when `b + a = c`. -/
+noncomputable def comp {a b c : M} (f : ShiftedHom X Y a) (g : ShiftedHom Y Z b) (h : b + a = c) :
+    ShiftedHom X Z c :=
+  f ≫ g⟦a⟧' ≫ (shiftFunctorAdd' C b a c h).inv.app _
+
+lemma comp_assoc {a₁ a₂ a₃ a₁₂ a₂₃ a : M}
+    (α : ShiftedHom X Y a₁) (β : ShiftedHom Y Z a₂) (γ : ShiftedHom Z T a₃)
+    (h₁₂ : a₂ + a₁ = a₁₂) (h₂₃ : a₃ + a₂ = a₂₃) (h : a₃ + a₂ + a₁ = a) :
+    (α.comp β h₁₂).comp γ (show a₃ + a₁₂ = a by rw [← h₁₂, ← add_assoc, h]) =
+      α.comp (β.comp γ h₂₃) (by rw [← h₂₃, h]) := by
+  simp only [comp, assoc, Functor.map_comp,
+    shiftFunctorAdd'_assoc_inv_app a₃ a₂ a₁ a₂₃ a₁₂ a h₂₃ h₁₂ h,
+    ← NatTrans.naturality_assoc, Functor.comp_map]
+
+/-! In degree `0 : M`, shifted hom `ShiftedHom X Y 0` identify to morphisms `X ⟶ Y`.
+We generalize this to `m₀ : M` such that `m₀ : 0` as it shall be convenient when we
+apply this with `M := ℤ` and `m₀` the coercion of `0 : ℕ`. -/
+
+/-- The element of `ShiftedHom X Y m₀` (when `m₀ = 0`) attached to a morphism `X ⟶ Y`. -/
+noncomputable def mk₀ (m₀ : M) (hm₀ : m₀ = 0) (f : X ⟶ Y) : ShiftedHom X Y m₀ :=
+  f ≫ (shiftFunctorZero' C m₀ hm₀).inv.app Y
+
+/-- The bijection `(X ⟶ Y) ≃ ShiftedHom X Y m₀` when `m₀ = 0`. -/
+@[simps apply]
+noncomputable def homEquiv (m₀ : M) (hm₀ : m₀ = 0) : (X ⟶ Y) ≃ ShiftedHom X Y m₀ where
+  toFun f := mk₀ m₀ hm₀ f
+  invFun g := g ≫ (shiftFunctorZero' C m₀ hm₀).hom.app Y
+  left_inv f := by simp [mk₀]
+  right_inv g := by simp [mk₀]
+
+lemma mk₀_comp (m₀ : M) (hm₀ : m₀ = 0) (f : X ⟶ Y) {a : M} (g : ShiftedHom Y Z a) :
```

Comment:
Should this be a simp lemma?

Reply:
I would rather not make these simp lemmas. If I had defined `ShiftedHom` as a single field `hom` structure, it would have very much made sense to have such a simp(s) lemma `mk₀_comp_hom`, but otherwise, I prefer not to. (I will use a single field structure when I repackage this for Ext-groups #13599.)

Reply:
Fair enough.